atdgen/src/oj_emit.ml

```diff
@@ -124,120 +118,44 @@ let get_assoc_type deref loc x =
-    }::acc
-  ) (0,[]) (Array.mapi (fun i x -> (i, x)) a) in
-
-  let fc = List.length acc in
-  let fm = Hashtbl.create fc in
-  let jfdir = Hashtbl.create fc in
-  let neg_one = List.fold_left (fun n f ->
-    Hashtbl.replace fm n f;
-    Hashtbl.replace jfdir f.jsonf.Json.json_fname n;
-    n - 1
-  ) (fc - 1) acc in
-  assert (neg_one = -1);
-
-  let existing_constr constr =
-    try Some (Hashtbl.find jfdir constr)
-    with Not_found -> None
-  in
-  (* Add implicit fields and index the deconstructed/tag field relations *)
-  (* TODO why is a fold only being used for its side effect? *)
-  let (_ : int) = Hashtbl.fold (fun i { jsonf = {Json.json_tag_field; _}; _ } k ->
-    match json_tag_field with
-    | None -> k
-    | Some constr ->
-        let field = Hashtbl.find fm i in
-        match existing_constr constr with
-        | Some c_i ->
-            let consf = Hashtbl.find fm c_i in
-            Hashtbl.replace fm i   { field with constructor = Some c_i };
-            Hashtbl.replace fm c_i { consf with payloads = i::consf.payloads };
-            k
-        | None -> (* Synthesize implicit field *)
-            let c_i = Hashtbl.length fm in
-            let f_name = implicit_field_name constr in
-            let ocamlf = {
-              Ocaml.ocaml_fname = f_name;
-              ocaml_default = None;
-              ocaml_mutable = false;
-              ocaml_fdoc = None;
-            } in
-            let jsonf = {
-              Json.json_fname = constr;
-              json_tag_field = None;
-              json_unwrapped = false;
-            } in
-            let synloc = (Lexing.dummy_pos, Lexing.dummy_pos) in
-            let mapping = {
-              f_loc = synloc;
-              f_name = f_name;
-              f_kind = Required;
-              f_value = String (synloc, Ocaml.Repr.String, Json.String);
-              f_arepr = Ocaml.Repr.Field ocamlf;
-              f_brepr = Json.Field jsonf;
-            } in
-            let imp = {
-              mapping = mapping;
-              default = Checked k;
-              ocamlf = ocamlf;
-              jsonf = jsonf;
-              field_ref = "field_"^f_name;
-              constructor = None;
-              payloads = [i];
-              implicit = true;
-            } in
-            Hashtbl.replace fm i   { field with constructor = Some c_i };
-            Hashtbl.replace fm c_i imp;
-            Hashtbl.replace jfdir constr c_i;
-            k
-  ) (Hashtbl.copy fm) k
-  in
-  let a = Array.make (Hashtbl.length fm) (Hashtbl.find fm 0) in
-  Array.iteri (fun n _ -> a.(n) <- Hashtbl.find fm n) a;
-  a
+  List.map (
+    fun x ->
+      let ocaml_fname, ocaml_default, json_fname, optional, unwrapped =
+        match x.f_arepr, x.f_brepr with
+        | Ocaml.Repr.Field o, Json.Field j ->
+            let ocaml_default =
+              match x.f_kind with
+                With_default ->
+                  (match o.Ocaml.ocaml_default with
```

Comment:
It's very likely because I undid features by restoring old code from before those features. I'll take a look at your cleanup commits and try to re-apply them.

Reply:
ok, I hand-restored your improvements to the `get_fields` function.